### DISCUSSION TITLE
Berry: Virtual switch definition in driver

### DISCUSSION THREAD

**Discussion by msetina** [arendst/Tasmota#19611] [Q&A] [answered]
Is there a way for Berry driver to present to tasmota data that would tell it to present a virtual switch? 
I am looking for a way to implement a boolean attribute in berry driver that HomeAssistant would present as a switch.

**Comment by sfromis** [comment]
Assuming Tasmota terms, not Home Assistant terms, I think that you are talking about a Relay *output*, not a Switch *input*. Right?

You have two ways of creating virtual relays in Tasmota. One is to config an unused gpio as a Relay pin.
In Berry, you additionally can [set](https://tasmota.github.io/docs/Berry/#low-level-access-to-tasmota-globals-and-settings)
`tasmota.global.devices_present` - Number of Power channels, e.g. having virtual relays

After that, you have the usual Berry mechanisms for power channels, [including](https://tasmota.github.io/docs/Berry/#lights-and-relays) but not limited to `tasmota.get_power()`, `tasmota.set_power()` and `tasmota.set_power_handler()`

**Reply by msetina** [reply]
Yes. Thank you for the guidance, I did not think of the semantics and you are correct from user point of view it is a relay.
Digging from your ling I found
https://github.com/arendst/Tasmota/blob/development/tasmota/berry/drivers/Shift595.be
And saw that it implemented 
def set_power_handler(cmd, idx)
which is probably called by tasmota for each driver present. Is it true?
How does get_power pathway work? Is there a handler to implement?
In my case the relay state could change internaly. Is it better to use set_power() so tasmota saves state of the relay?

**Reply by sfromis** [reply, marked as answer]
Yes, when a driver is registered, the callback `set_power_handler` gets notified when some power channel change happen, and the idx value is the combination of channels currenly on.

`tasmota.get_power` is a function you call to get the status of one or all power channels. No handler or pathway other than issuing the call.

When asking about "better?", no clear answers without details, like compared to? If the Berry code decides to turn on/off a power channel, `tasmota.set_power` is the most direct way of doing that, instead of issuing a command via `tasmota.set_power`. But you did not explain what you mean by "the relay state could change *internally*", and in what context.

Of course, you can also flip a gpio output without characterizing it as a power channel, and just skip the usual Tasmota logic about messages and saving around power channel changes.

**Reply by msetina** [reply]
My "question of better" is fishing for suggestion not knowing where the state of a relay is managed. If only set_power pathway exists with set_power_handler the state of power/relay with no physical state to check at get_power call has to depend on something. If driver implementation does not reply to some sort of inquiry, tasmota needs to manage states with that it "sees" when set_power is called.
By internaly I refered to a bool state driver has and can change without calling set_power. In that case states may differ with what tasmota   has seen from set_power calls.
So the correct (not better in my words) to protect tasmota states state is to manipulate and use set_power/get_power of tasmota to also be sure that tasmota reports synced state to the outside world.

**Reply by sfromis** [reply]
Basically, `tasmota.get_power` can be seen as the closest you can get to "physical state", unless you do something at the hardware level, and from Berry other ways than using `tasmota.set_power` would be "less direct". In principle, I'd suggest to not have your driver even keep a bool with a state to keep synchronized, unless you have good reasons for the redundancy. And then use `tasmota.get_power` when you need to act depending on current state.

But still, it all depends on details of what you want to achieve, and what level of functionality is "enough". In most cases, having the full Tasmota "circus" around maintaining and reporting of power channel state changes make good sense. And sometimes, less may be fine for the purpose.

**Reply by msetina** [reply]
Thank you

**Reply by msetina** [reply]
Little feedback: Current HomeAssistant integration for tasmota takes additional relays and shows them as switches. Also it does not show the Friendly names. Instead it shows switch_9 instead of relay_9 which is a little confusing

**Reply by sfromis** [reply]
Tasmota only provides 8 "friendly names", hence nothing such for > 8. I'm not using HA, but AFAIK, they've chosen to call a relay (or other "power" channels) a switch instead.

**Reply by msetina** [reply]
In tasmota I could add FriendlyName and WebButton name for all the virtual relays Berry driver adds. Only the HA integration does not take them. In JSON they are POWER9 - POWER13

**Reply by sfromis** [reply]
Standard tasmota does only provide for 8 FriendlyName, but up to 32 WebButton.

**Reply by msetina** [reply]
Looking at the discovery MQTT JSON I see that there is place for only 8 FriendlyNames. Is there any plans to extend that for tasmota32 as there is a posibility to prepare more relays with ESP32. Especialy as the HA integration uses FriendlyName on presentation side. On the other hand WebButton names could be something to add to discovery. Web is a presentation layer and HA as presentation layer could be interested in those name.

**Reply by barbudor** [reply]
> Web is a presentation layer and HA as presentation layer could be interested in those name

Any home automation system is free to ask Tasmota whatever it wants to know using the MQTT commands. The main purpose of the discovery is to say "hello, here I am", not to dump the whole configuration

**Reply by sfromis** [reply]
Tasmota only **has** 8 "FriendlyName". If more were provided for, they'd be in the discovery payloads too. And yes, you can ask for the `WebButton` names by publishing that command via MQTT.
